pkg/util/heap/heap.go

```diff
@@ -0,0 +1,211 @@
+		obj:   keyValue.obj,
+		index: len(h.keys),
+	}
+	h.keys = append(h.keys, keyValue.key)
+}
+
+// Pop is supposed to be called by heap.Pop only.
+func (h *data) Pop() interface{} {
+	key := h.keys[len(h.keys)-1]
+	h.keys = h.keys[:len(h.keys)-1]
+	item, ok := h.items[key]
+	if !ok {
+		// This is an error
+		return nil
+	}
+	delete(h.items, key)
+	return item.obj
+}
+
+// Peek is supposed to return the head of the heap
+func (h *data) Peek() interface{} {
+	if len(h.keys) > 0 {
+		return h.items[h.keys[0]].obj
+	}
+	return nil
+}
+
+// Heap is a producer/consumer queue that implements a heap data structure.
+// It can be used to implement priority queues and similar data structures.
+type Heap struct {
+	data *data
+}
+
+// AddOrUpdate inserts an item to the queue.
+// The item will be updated if it already exists.
+func (h *Heap) AddOrUpdate(obj interface{}) error {
+	key := h.data.keyFunc(obj)
+	if _, exists := h.data.items[key]; exists {
+		h.data.items[key].obj = obj
+		heap.Fix(h.data, h.data.items[key].index)
+	} else {
+		heap.Push(h.data, &itemKeyValue{key, obj})
+	}
+
+	return nil
+}
+
+// AddIfNotPresent inserts an item to the queue. If an item with
+// the key is present in the map, no changes is made to the item.
+func (h *Heap) AddIfNotPresent(obj interface{}) (added bool, err error) {
+	key := h.data.keyFunc(obj)
+	if _, exists := h.data.items[key]; exists {
+		return false, nil
+	}
+
+	heap.Push(h.data, &itemKeyValue{key, obj})
+	return true, nil
+}
+
+// Delete removes an item.
+func (h *Heap) Delete(obj interface{}) error {
+	key := h.data.keyFunc(obj)
+	item, ok := h.data.items[key]
+	if !ok {
+		return fmt.Errorf("object not found")
+	}
+
+	heap.Remove(h.data, item.index)
+	return nil
+}
+
+// Pop returns the head of the heap and removes it.
+func (h *Heap) Pop() (interface{}, error) {
+	obj := heap.Pop(h.data)
+	if obj == nil {
+		return nil, fmt.Errorf("object was already removed from heap data")
+	}
+	return obj, nil
+}
+
+// Peek returns the head of the heap without removing it.
+func (h *Heap) Peek() interface{} {
+	return h.data.Peek()
+}
+
+// Get returns the requested item, exists, error.
+func (h *Heap) Get(obj interface{}) (item interface{}, exists bool) {
+	key := h.data.keyFunc(obj)
+	return h.GetByKey(key)
+}
+
+// GetByKey returns the requested item, or sets exists=false.
+func (h *Heap) GetByKey(key string) (interface{}, bool) {
+	item, exists := h.data.items[key]
+	if !exists {
+		return nil, false
+	}
+	return item.obj, true
+}
+
+// Len returns the number of items in the heap.
+func (h *Heap) Len() int {
+	return len(h.data.keys)
+}
+
+// List returns a list of all the items.
+func (h *Heap) List() []interface{} {
+	list := make([]interface{}, 0, h.Len())
+	for _, item := range h.data.items {
+		list = append(list, item.obj)
+	}
+	return list
+}
+
+// New returns a Heap which can be used to queue up items to process.
+func New(keyFn keyFunc, lessFn lessFunc) *Heap {
+	return &Heap{
+		data: &data{
```

Comment:
does it need to be a pointer?

Reply:
nope, but the operation of native heap is always need a pointer, like `heap.Fix(&data, i)`, so I prefer pointer here.

Reply:
similar to the argument below, avoid using pointers if not needed for performance reasons. There is no issue with doing `heap.Fix(&data, i)`

Reply:
Updated.